docs/source/clickhouse/schema_design.rst

```diff
@@ -11,36 +21,64 @@ and not load them for every row for every query is part of the performance advan
 ClickHouse provides over a traditional RDBMS (like PostgreSQL).
 
 Commonly, a data schema contains a flexible key:value pair mapping
-(canonically: ``tags`` or ``contexts``) and stores that
+(canonically at Sentry: ``tags`` or ``contexts``) and stores that
 data in a column that contains two ``Nested`` arrays where the first array contains the keys
-of the dictionary and the second array contains the values. This works well when
-your dataset design gives you the ability to filter based on other attributes to a small
+of the dictionary and the second array contains the values, indexed by a bloom filter. This works well when
+your dataset and query design gives you the ability to filter for exact matches to a small
 number of rows and the flexibility of completely arbitrary keys and values are a real requirement.
-Often, however, a ClickHouse user is interested in rows that contain a specific tag key or a
-specific ``key=value`` pair. In this case one is often best-served by creating a top-level
-column and "promoting" the data to it [#dupe]_.
+Often, however, a ClickHouse query filters for rows that contain a substring match or regular
+expression match for a tag value of a given key. This makes bloom filter (see :ref:`bloom`) indexes
+not usable for the query and requires moving (or promoting) those relevant values for a given tag to
+a new separate column which can support these queries more efficiently [#dupe]_.
+
+It is very efficient for queries to filter by a specific matching value that exists in
+a low number of rows for a given tag key being non-empty or key:value pair on a column
+that contains **just** the values of that specific key. However, in :ref:`selectivity`
+we discuss why this fails when queries contain conditions for a specific ``key=value``
+pair that occurs in a very large percentage of the table data (for example,
+``http_status_code=200`` or ``hasKey(http_status_code)``).
+
+.. _selectivity:
 
-This promotion is preferable on a couple dimensions:
+Selectivity in queries and indices
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-1. Even though it looks like a dictionary, our tags and contexts "maps" are actually arrays of keys and values.
-   With array accesses Clickhouse needs to locate the index of the key whose value needs to be looked up.
-   That is an O(N) lookup. The same key may be located at different indexes for each row.
-   Once the index is known, looking up its value is O(1) since you perform direct index access.
-   If the query has to lookup M rows, this becomes an O(M * N) in terms of time complexity.
+Queries are much more efficient when they have the attribute of being low-selectivity on
+the table indexes -- meaning the query conditions on indexed columns filter the dataset
+to a very small proportion of the overall number of rows. Therefore it's ideal to design
+schemas and queries that cause searches on a very low number of rows. High selectivity
```

Comment:
I think this may be a little misleading. 
The point of a real time analytics database is to be able to produce aggregations over very large amount of rows very fast. This is the primary use case and Clickhouse does great at that.
What is written here make it seems that such usage is an anti-pattern. If we had a use case where most queries were very low selectivity Clickhouse would not be the right storage. A key value pair would be the one. 

The point of introducing these optimization is to make low selectivity queries still similarily efficient as they would be on a database with b-tree indexes like postgres by reducing the amount of data to scan.

Reply:
I agree, this is missing some nuance. It depends on the type of conditions used. A `count()` over 90 days with a simple `project_id` filter can execute much faster than a regex search on tags over 7 days, even though the second is much lower selectivity.

Reply:
Rather than try to address the nuance Evan is suggesting, I think I'm just going to drop that wording and say "lower selectivity = better".

We note in the earlier and later sections that bloom filters can't do anything with fuzzy (regex/LIKE) filtering.